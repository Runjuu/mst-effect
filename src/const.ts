export const Identity: unique symbol = Symbol.for('mst-effect-identity')

export const HANDLE_MST_EFFECT_ACTIONS = 'HANDLE_MST_EFFECT_ACTIONS'
