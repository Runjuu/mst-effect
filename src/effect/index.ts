export * from './effect'
export { action, ValidEffectActions, EffectAction } from './action'
