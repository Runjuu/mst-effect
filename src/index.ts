export * from './types'
export * from './effect'
export * from './mst'
