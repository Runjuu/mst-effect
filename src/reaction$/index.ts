import { reaction, IReactionPublic, IReactionOptions } from 'mobx'
import { Observable } from 'rxjs'

export function reaction$<T>(
  expression: (r: IReactionPublic) => T,
  opts: IReactionOptions<T, true>,
): Observable<{ current: T; prev: T | undefined; r: IReactionPublic }>

export function reaction$<T>(
  expression: (r: IReactionPublic) => T,
  opts?: IReactionOptions<T, false>,
): Observable<{ current: T; prev: T; r: IReactionPublic }>

export function reaction$<T, FireImmediately extends boolean = false>(
  expression: (r: IReactionPublic) => T,
  opts?: IReactionOptions<T, FireImmediately>,
): Observable<{ current: T; prev?: T; r: IReactionPublic }> {
  return new Observable((observer) => {
    const dispose = reaction(
      expression,
      (current, prev, r) => observer.next({ current, prev, r }),
      opts,
    )

    return () => dispose()
  })
}
