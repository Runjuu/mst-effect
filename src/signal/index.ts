import { addDisposer } from 'mobx-state-tree'
import { Observable, Subject } from 'rxjs'

import { AnyInstance, PayloadFunc } from '../types'

export type SignalFactory<P, R> = (payload$: Observable<P>) => Observable<R>

export type SignalResult<P, R> = [Observable<R>, PayloadFunc<P, void>]

export function signal<P>(self: AnyInstance): SignalResult<P, P>
export function signal<P, R>(self: AnyInstance, fn: SignalFactory<P, R>): SignalResult<P, R>
export function signal<P, R>(self: AnyInstance, fn?: SignalFactory<P, R>): SignalResult<P, R> {
  const payloadSource = new Subject<P>()
  const signal$ = fn?.(payloadSource.asObservable()) ?? payloadSource.asObservable()

  addDisposer(self, () => payloadSource.complete())

  return [signal$, (payload: P) => payloadSource.next(payload)] as SignalResult<P, R>
}
