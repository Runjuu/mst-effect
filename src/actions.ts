import { EffectAction, ValidEffectActions } from './types'
import { Identity } from './const'

export function action<P extends any[]>(fn: (...params: P) => void, ...params: P): EffectAction {
  return { [Identity]: () => fn(...params) }
}

export function noop(..._params: any[]): EffectAction {
  return action(() => {})
}

export function runActions(actions: ValidEffectActions): void {
  if (Array.isArray(actions)) {
    actions.forEach(runAction)
  } else {
    runAction(actions)
  }
}

function isValidAction(action: any): action is EffectAction {
  return action && typeof action === 'object' && typeof action[Identity] === 'function'
}

function runAction(action: any): void {
  if (isValidAction(action)) {
    action[Identity]()
  } else if (action !== null) {
    console.warn(`[mst-effect]: ${action} is not a valid EffectActions`)
  }
}
