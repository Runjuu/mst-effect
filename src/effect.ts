import { Subject } from 'rxjs'
import { tap, catchError } from 'rxjs/operators'
import { IAnyModelType, Instance, addDisposer } from 'mobx-state-tree'

import type { PayloadFunc, EffectFactory } from './types'
import { HANDLE_MST_EFFECT_ACTIONS } from './const'

export function effect<P = void>(
  self: Instance<IAnyModelType>,
  fn: EffectFactory<P>,
): PayloadFunc<P, void>
export function effect(
  self: Instance<IAnyModelType>,
  fn: EffectFactory<any>,
): (payload: any) => void {
  const payloadSource = new Subject()
  const effect$ = fn(payloadSource.asObservable())
  const subscription = effect$
    .pipe(
      tap((actions) => {
        if (self[HANDLE_MST_EFFECT_ACTIONS]) {
          self[HANDLE_MST_EFFECT_ACTIONS](actions)
        } else {
          console.warn(
            `[mst-effect]: Make sure the 'types' is imported from 'mst-effect' instead of 'mobx-state-tree'`,
          )
        }
      }),
      logAngIgnoreError(fn),
    )
    .subscribe()

  addDisposer(self, () => {
    payloadSource.complete()
    subscription.unsubscribe()
  })

  return (payload) => {
    payloadSource.next(payload)
  }
}

function logAngIgnoreError(factory: unknown) {
  return catchError((err, caught) => {
    /* eslint-disable no-console */
    console.group('[mst-effect]: error')
    console.log(factory)
    console.error(err)
    console.groupEnd()
    /* eslint-enable no-console */
    return caught
  })
}
