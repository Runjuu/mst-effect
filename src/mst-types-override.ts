import { types as mstTypes } from 'mobx-state-tree'

import { HANDLE_MST_EFFECT_ACTIONS } from './const'
import { runActions } from './actions'

export const types: typeof mstTypes = {
  ...mstTypes,
  model(...params: any[]) {
    return mstTypes.model(...params).actions(() => ({
      [HANDLE_MST_EFFECT_ACTIONS]: runActions,
    }))
  },
}
