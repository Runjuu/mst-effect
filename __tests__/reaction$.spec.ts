import { Observable } from 'rxjs'
import { expectType } from 'tsd'
import { IReactionPublic } from 'mobx'

import { types, reaction$ } from '../src'

const Model = types
  .model({
    num: types.number,
  })
  .actions((self) => ({
    updateNum(num: number) {
      self.num = num
    },
  }))

describe(`reaction$`, () => {
  it(`should return an Observable`, () => {
    expect(reaction$(() => {})).toBeInstanceOf(Observable)
  })

  it(`should emit data anytime it observes changes`, () => {
    const model = Model.create({ num: 0 })
    const num$ = reaction$(() => model.num)
    const spy = jest.fn()

    num$.subscribe(({ current, prev }) => spy({ current, prev }))

    model.updateNum(1)
    model.updateNum(2)
    model.updateNum(3)

    expect(spy).toBeCalledTimes(3)
    expect(spy.mock.calls).toMatchSnapshot()
  })

  it(`should un-observe after unsubscribe`, () => {
    const model = Model.create({ num: 0 })
    const spy = jest.fn()
    const num$ = reaction$(() => {
      spy(model.num)
      return model.num
    })

    num$.subscribe().unsubscribe()
    model.updateNum(1)
    model.updateNum(2)
    model.updateNum(3)

    expect(spy).toBeCalledTimes(1)
    expect(spy).toBeCalledWith(0)
  })

  it(`should able to specify reaction's options`, () => {
    const model = Model.create({ num: 0 })
    const num$ = reaction$(() => model.num, { fireImmediately: true })
    const spy = jest.fn()

    num$.subscribe(({ current, prev }) => spy({ current, prev }))

    expect(spy).toBeCalledTimes(1)
    expect(spy.mock.calls).toMatchSnapshot()
  })

  it(`should mark 'prev' to optional when 'fireImmediately' is true`, () => {
    const model = Model.create({ num: 0 })
    const num$ = reaction$(() => model.num, { fireImmediately: true })
    expectType<Observable<{ current: number; prev?: number; r: IReactionPublic }>>(num$)
  })
})
