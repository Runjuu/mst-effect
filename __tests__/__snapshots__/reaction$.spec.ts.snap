// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reaction$ should able to specify reaction's options 1`] = `
Array [
  Array [
    Object {
      "current": 0,
      "prev": undefined,
    },
  ],
]
`;

exports[`reaction$ should emit data anytime it observes changes 1`] = `
Array [
  Array [
    Object {
      "current": 1,
      "prev": 0,
    },
  ],
  Array [
    Object {
      "current": 2,
      "prev": 1,
    },
  ],
  Array [
    Object {
      "current": 3,
      "prev": 2,
    },
  ],
]
`;
